### ISSUE TITLE
Add configuration options to `explicit_string_variable`

### ISSUE THREAD

**Issue by haggys22** [FriendsOfPHP/PHP-CS-Fixer#4332]
**PHP version:** `PHP 7.2.14`

**PHP CS Fixer version:** `PHP CS Fixer 2.14.0`

**Command used to run PHP CS Fixer:** `php-cs-fixer fix test.php --rules=explicit_string_variable`

This issue is related to #3488 but with a different scope. I've also mentioned this on Gitter https://gitter.im/PHP-CS-Fixer/Lobby?at=5c8686b5d3b35423cba03d52

I appreciate the default rule to use the "simple" `${` (T_DOLLAR_OPEN_CURLY_BRACES) syntax over the "complex" `{$` (T_CURLY_OPEN) one (although the PHP documentation states ["This isn't called complex because the syntax is complex, but because it allows for the use of complex expressions."](https://secure.php.net/manual/en/language.types.string.php#language.types.string.parsing.complex))

Currently this rule can lead to inconsistencies as it does not enforce the simple sytax and in special when both, simple and complex, expressions are used in the same project:

```
-echo "$foo ${bar} {$baz} {$something->more['complex']}";
+echo "${foo} ${bar} {$baz} {$something->more['complex']}";
```

Therefore I propose to add configuration options
1. for the syntax to be used for simple expressions
2. if the configured syntax should be enforced (i.e. already explicit string variables using the other syntax should be fixed to the configured one)

```
'explicit_string_variable' => false|[
    'syntax' => 'simple' (default)|'complex',
    'enforce' => false (default)|true
]|true (use defaults, i.e. the current behavior)
```

**Expected behavior:**

* before running PHP CS Fixer (no changes):
```
echo "$foo ${bar} {$baz} {$something->more['complex']}";
```

* with `'syntax' => 'simple', 'enforce' => false` (default, current behavior):
```
echo "${foo} ${bar} {$baz} {$something->more['complex']}";
```

* with `'syntax' => 'simple', 'enforce' => true`:
```
echo "${foo} ${bar} ${baz} {$something->more['complex']}";
```

* with `'syntax' => 'complex', 'enforce' => false`:
```
echo "{$foo} ${bar} {$baz} {$something->more['complex']}";
```

* with `'syntax' => 'complex', 'enforce' => true`:
```
echo "{$foo} {$bar} {$baz} {$something->more['complex']}";
```

**Comment by SpacePossum** [comment]
Hi @haggys22 and thanks for your suggestion!

I'm not in favor of adding the `'force'` flag, I think the rule should always fix where possible and leave other constructs untouched. I feel this is consistent to how most of the rules work. Secondly, defining the case where `'force'`  should apply might become complex or unclear (apply to one line, or the whole file, or namespace etc.)

The other `'syntax'` option makes sense to me, as there is no real standard for the notation AFAIK, so supporting both might be nice for users.

**Comment by haggys22** [comment]
@SpacePossum I think to be consistent with other rules, `'explicit_string_variable'` should always enforce the use of one syntax and the current behavior, using simple syntax, should already result in

```
echo "${foo} ${bar} ${baz} {$something->more['complex']}";
```

The `'force'` option was just an idea to be able to maintain the current behavior.

@dmvdbrugge I have to admit I only looked for issues, not PRs, addressing this.

For the coding style I want to achieve (i.e. explicit complex string variables for simple expressions), `explicit_string_variable` combined with `simple_to_complex_string_variables` would be sufficient.

```
echo "${foo} ${bar} {$baz} {$something->more['complex']}"; // after 'explicit_string_variable'
echo "{$foo} {$bar} {$baz} {$something->more['complex']}"; // after 'simple_to_complex_string_variables'
```

I'm not sure if this should be considered, but it would not be possible to enforce simple syntax for simple expressions (i.e. fix simple expressions using complex syntax to simple syntax)

```
echo "${foo} ${bar} {$baz} {$something->more['complex']}"; // after 'explicit_string_variable'
echo "${foo} ${bar} ${baz} {$something->more['complex']}"; // 'complex_to_simple_string_variables' does not exist
```

**Comment by dmvdbrugge** [comment]
I started the discussion about this in June 2018 on Gitter [here](https://gitter.im/PHP-CS-Fixer/Lobby?at=5b114b9d352b9e1a4b4e2ec8). I offered the option approach but was corrected into doing it as new fixer. To quote keradus:

> the thing is that if you do it as an option @dmvdbrugge , it would only follow that `{$` for code introduced by that rule, and not for already existing `${`. to have `${`->`{$` applied globally, we need standalone rule anyway. and there is no need to make same code twice, as new rule and option for existing rule

Discussion continued [1,5 month later](https://gitter.im/PHP-CS-Fixer/Lobby?at=5b56447cf10ffa0b48a90166) when I created the PR. Your question:

> I'm not sure if this should be considered, but it would not be possible to enforce simple syntax for simple expressions (i.e. fix simple expressions using complex syntax to simple syntax)

Is answered there:

> Because from explicit to implicit seems like a bad choice, and we wouldn't want to recommend it, and going from `{$` back to `${` in my opinion only removes consistency (which is why I created it in the first place)

**Comment by dmvdbrugge** [comment]
@keradus I think this issue can be closed, as `simple_to_complex_string_variable` has been available since 2.15 and they said

> For the coding style I want to achieve (i.e. explicit complex string variables for simple expressions), `explicit_string_variable` combined with `simple_to_complex_string_variables` would be sufficient.

And all other arguments have been refuted.

**Comment by SpacePossum** [comment]
closing as https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/3942 has been merged, thanks @dmvdbrugge !  :+1:
